---
title: "MA_Project1"
author: "Pia Chouaifaty"
date: "10/12/2020"
output: html_document
---

```{r include=FALSE}
knitr::opts_chunk$set(fig.width=8, fig.height=8)
```

# Project Description

### Epithelial-mesenchymal status renders differential responses to cisplatin in ovarian cancer

The data set used is from a paper studying the differences in gene expression in mesenchymal and epithelial cells following cisplatin treatment. 171 samples of various cell lines treated with cisplatin and negative controls were obtained. The aim is to analyze differential expression as a response to cisplatin treatment between mesenchymal and epithelial cells.

Citation:  
Miow, Q., Tan, T., Ye, J. et al. Epithelial–mesenchymal status renders differential responses to cisplatin in ovarian cancer. Oncogene 34, 1899–1907 (2015). https://doi.org/10.1038/onc.2014.136

#Pipeline

## 1 Libraries

```{r libraries, echo=TRUE, message=FALSE, warning=FALSE}

#General Bioconductor packages
library(Biobase)
library(oligoClasses)
     
#Annotation and data import packages
library(ArrayExpress)
library(pd.hugene.1.0.st.v1)
library(hugene10sttranscriptcluster.db)
     
#Quality control and pre-processing packages
library(oligo)
library(arrayQualityMetrics)
     
#Analysis and statistics packages
library(limma)
library(topGO)
library(ReactomePA)
library(clusterProfiler)
     
#Plotting and color options packages
library(gplots)
library(ggplot2)
library(geneplotter)
library(RColorBrewer)
library(pheatmap)
     
#Formatting/documentation packages
   #library(rmarkdown)
   #library(BiocStyle)
library(dplyr)
library(tidyr)

#Helpers:
library(stringr)
library(matrixStats)
library(genefilter)
library(openxlsx)
   #library(devtools)

```

## 2 Downloading the Raw & Processed Data From Array Express

[ArrayExpress Experiment Link](https://www.ebi.ac.uk/arrayexpress/experiments/E-GEOD-47856/)  
Accession number: E-GEOD-47856

```{r}
raw_data_dir = "/Users/piachouaifaty/bif425_fall20_chouaifaty_pia_copy/MA_Pipeline_Project1/raw_data_files"
if(!dir.exists(raw_data_dir)) {dir.create(raw_data_dir)}

#anno_AE = getAE("E-GEOD-47856", path=raw_data_dir, type="raw")
```


```{r}
processed_data_dir = "/Users/piachouaifaty/bif425_fall20_chouaifaty_pia_copy/MA_Pipeline_Project1/processed_data"
if(!dir.exists(processed_data_dir)) {dir.create(processed_data_dir)}

#anno_AE = getAE("E-GEOD-47856", path=processed_data_dir, type="processed")
```


```{r}
sdrf_location = file.path(raw_data_dir, "E-GEOD-47856.sdrf.txt")
SDRF = read.delim(sdrf_location)
rownames(SDRF) = SDRF$Array.Data.File
```

## 3 Appending EMT Classification to SDRF

The authors supplied a list of cell lines that they found to be epithelial-like/mesenchymal-like in an additional experiment in the supplementary material.  
In order to make it easier to separate based on cell line EMT status (mesenchymal vs epithelial) later on, I add a column to the SDRF matrix specifying EMT status before transforming it into an annotated data frame.  

The sample_classification.xls file was extracted from the supplementary data provided by the paper. sample_class is a dataframe containing the sample cell line and its classification as either epithelial-like or mesenchymal-like.

```{r}
#install.packages("readxl")
library(readxl)

suppl_dir = "/Users/piachouaifaty/bif425_fall20_chouaifaty_pia_copy/MA_Pipeline_Project1"
EMT_location = file.path(suppl_dir, "sample_classification.xls")
EMT_status = read_excel(EMT_location)
```

I now add the Classification column to the SDRF data frame.

```{r}
SDRF = left_join(SDRF, EMT_status, by = c("FactorValue..CELL.LINE." = "Sample"))
rownames(SDRF) = SDRF$Array.Data.File
SDRF2 = SDRF #for reference
rownames(SDRF2) = SDRF2$Array.Data.File
SDRF = AnnotatedDataFrame(SDRF)
```

Creating an ExpressionSet for the raw data

```{r}
raw_data = oligo::read.celfiles(filenames = file.path(raw_data_dir, SDRF$Array.Data.File), verbose = FALSE, phenoData = SDRF)
stopifnot(validObject(raw_data))
```

Taking a look at the raw data

```{r}
head(Biobase::pData(raw_data))
```

Retain only columns that we need:  
1. Sample identifier (Source.Name)  
2. Treatment or control (FactorValue..TREATMENT)  
3. Cell line name (FactorValue..CELL.LINE.)
4. Cell line EMT classification (Classification)  

```{r}
Biobase::pData(raw_data) = Biobase::pData(raw_data)[,c("Source.Name", "FactorValue..TREATMENT.", "FactorValue..CELL.LINE.", "Classification")]
SDRF2 = SDRF2[,c("Source.Name", "FactorValue..TREATMENT.", "FactorValue..CELL.LINE.","Classification")]

head(Biobase::pData(raw_data))
```

## 4 Log-Transformation & Visualization

We take the log2 of the data, and plot the PCA to check if we have clustering based on cell line classification and/or treatment (cisplatin vs control).

```{r}
exp_raw = log2(Biobase::exprs(raw_data))
```

PCA Analysis & Plot

```{r}
PCA_raw <- prcomp(t(exp_raw), scale. = FALSE)

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    EMT = pData(raw_data)$Classification,
                    Treatment = pData(raw_data)$FactorValue..TREATMENT.,
                    Source = pData(raw_data)$Source.Name)

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = EMT, colour = Treatment)) +
  ggtitle("PCA plot of the log-transformed raw expression data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5))+
  coord_fixed(ratio = sd_ratio) +
  scale_shape_manual(values = c(4,15)) + 
  scale_color_manual(values = c("darkorange2", "dodgerblue4"))
```

There seems to be a clear clustering according to PC1 (48.7% variance, encompasses almost half the data). It seems that more epithelial-like cells are to the left while mesenchymal-like cells are to the right, but that will be clearer once we normalize and account for triplicate assays. 

Boxplot

```{r}
oligo::boxplot(raw_data, target="core", main = "Boxplot of log-2 intensities for the raw data")
```

Raw data report

```{r}
#arrayQualityMetrics(expressionset = raw_data,
    #outdir = raw_data_dir,
   # force = TRUE, do.logtransform = TRUE,
   # intgroup = c("FactorValue..TREATMENT.", "Classification"))
```

## 5 Background Adjustment, calibration, summarization and annotation

### RLE (Relative Log Expression) data quality analysis

#### Backround correction

```{r}
palmieri_eset = oligo::rma(raw_data, target="core", normalize=FALSE)
```

```{r}
head(Biobase::exprs(palmieri_eset))
```

#### Plotting the RLE

```{r}
row_medians_assayData <- 
  Biobase::rowMedians(as.matrix(Biobase::exprs(palmieri_eset)))

RLE_data <- sweep(Biobase::exprs(palmieri_eset), 1, row_medians_assayData)

RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- 
  tidyr::gather(RLE_data, patient_array, log2_expression_deviation)

#ylim is the ylimit of the x-axis

ggplot2::ggplot(RLE_data_gathered, aes(patient_array,
                                       log2_expression_deviation)) + 
  geom_boxplot(outlier.shape = NA) + 
  ylim(c(-2, 2)) + 
  theme(axis.text.x = element_text(colour = "aquamarine4", 
                                  angle = 60, size = 6.5, hjust = 1 ,
                                  face = "bold"))
```

The above plot shows the probe intensity per chip variation with respect to median intensities
We can see a couple of outliers to the left and a few to the right
Very large variances exist in the data (unexpected), however, the data is not normally distributed, because the authors used a u-test, so a boxplot probably isn't a good indicator anyway. The heatmap will probably be more useful.

#### RMA Calibration

In the original paper, they used the RMA algorithm to calibrate the data

```{r}
palmieri_eset_norm <- oligo::rma(raw_data, target = "core")
```


#### Comparing my normalized expression values with the original paper's

I join all the processed data into one matrix to easily compare it with my resulting expression values:

```{r}
for(row in 1:nrow(SDRF))
{processedfilenames=c(SDRF$Derived.Array.Data.File)}

samplepro = read.delim(file.path(processed_data_dir, processedfilenames[1]))
colnames(samplepro)=(c("ID_REF", processedfilenames[1]))

for(value in processedfilenames)
{
  processed_location = file.path(processed_data_dir, value)
  m=read.delim(processed_location)
  colnames(m)=(c("ID_REF", value))
  samplepro = left_join(samplepro, m)
}

rownames(samplepro)=c(samplepro$ID_REF)


head(samplepro)
```


```{r}
head(Biobase::exprs(palmieri_eset_norm))
```

So far the processed data from the original paper and the pipeline's data after normalization are very similar. However, there seems to be around a 1-unit difference [~0.8-1.20] between the two. In the paper, the authors stated having standardized the expression values using ComBat to remove the batch effect after running RMA, so that may be the cause of this discrepancy.

#### Heatmap clustering analysis

Sample-to-sample heatmap:

```{r}
#defining the phenotype names: if it includes "non", change to non_infl, otherwise
#change to inlf
EMT_st <- ifelse(str_detect(pData
                                    (palmieri_eset_norm)$Classification,
                             "Epithelial"), "epi", "mes")

treatment_names <- ifelse(str_detect(pData
                                    (palmieri_eset_norm)$FactorValue..TREATMENT.,
                             "none"), "none", "cis")

##annotation
annotation_for_heatmap <- 
  data.frame(EMT = EMT_st,  Treatment = treatment_names)

#setting rownames
row.names(annotation_for_heatmap) <- row.names(pData(palmieri_eset_norm))
```

```{r}
#the heatmap will calculate for each sample/individual/chip the distance with respect to other chips 

#the actual heatmap is generated here
# generate distances
dists <- as.matrix(dist(t(exp_palmieri), method = "manhattan"))
#manhattan |x1-x2|
#euclidian (x1-x2)^2

rownames(dists) <- row.names(pData(palmieri_eset_norm))
hmcol <- rev(colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255))
colnames(dists) <- NULL 
diag(dists) <- NA

#assigning colors for the conditions and diseases (annotation)
ann_colors <- list(
  EMT = c(epi = "chartreuse4", mes = "burlywood3"),
  Treatment = c(cis = "blue4", none = "cadetblue2")
                   )

#pheatmap: dists is the distance matrix that we calculated
pheatmap(dists, col = (hmcol), 
         annotation_row = annotation_for_heatmap,
         annotation_colors = ann_colors,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                         max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering heatmap for the calibrated samples")
```


#### PCA After Normalization

```{r}
exp_palmieri <- Biobase::exprs(palmieri_eset_norm) 
PCA <- prcomp(t(exp_palmieri), scale = FALSE)

percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    EMT = pData(raw_data)$Classification,
                    Treatment = pData(raw_data)$FactorValue..TREATMENT.,
                    Source = pData(raw_data)$Source.Name)


ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = EMT, colour = Treatment)) +
  ggtitle("PCA plot of the calibrated, summarized data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  coord_fixed(ratio = sd_ratio) +
  scale_shape_manual(values = c(4,15)) + 
  scale_color_manual(values = c("darkorange2", "dodgerblue4"))
```

#### Triplicate Values & Filtering Based on Variance

The cell lines for which there are more than 2 assays (cisplatin/control) are the following: A2008, A2780, C13, CH1, DOV13, DOV13B, FU-OV-1, HeyA8, HeyC2, IGROV-1, OV90, OVCA420, OVCA429, OVCA433, OVCAR-10, OVCAR-8, PA-1, TYK-nu, SKOV-3.  

The authors stated that they took advantage of the triplicate data by removing potentially fragile probes with strong variations (sd>0.2) within the triplicates from the data, decreasing the total probe number from 33297 to 21329. -> 11968 probes removed

After trying it several times, and because the paper was unclear on whether it used log-transformed values or not for this step, and also because of the additional step the authors performed after RMA, I decided to go with variance>0.28 as a cutoff to get a comparable number of probe removals and not remove too many probes that may be significant.

```{r}

triplicate_cell_lines = c("A2008", "A2780", "C13", "CH1", "DOV13", "DOV13B", "FU-OV-1", "HeyA8", "HeyC2", "IGROV-1", "OV90", "OVCA420", "OVCA429", "OVCA433", "OVCAR-10", "OVCAR-8", "PA-1", "TYK-nu")

probes = vector()

for(value in triplicate_cell_lines)
{cell_lines=subset(SDRF2, FactorValue..CELL.LINE. == value)

cisplatin=subset(cell_lines, FactorValue..TREATMENT. == "cisplatin")
cisn=row.names(cisplatin)

cis_df = exp_palmieri[,c(cisn)]
cisrown=row.names(cis_df)
for(r in 1:nrow(cis_df))
{
  stddev=sd(cis_df[r,])
  if ((stddev*stddev)>0.28)
    {probes=c(probes, cisrown[r])}
}

control=subset(cell_lines, FactorValue..TREATMENT. == "none")
contrn=row.names(control)

contr_df = exp_palmieri[,c(contrn)]
contrown=row.names(contr_df)
for(r in 1:nrow(contr_df))
{
  stddev=sd(cis_df[r,])
  if ((stddev*stddev)>0.28)
    {probes=c(probes, contrown[r])}
}}

#probes to be removed
probes=unique(probes)
```

Below is the number of probes that will be removed.

```{r}
length(probes)
probes
```

We then remove these probes from the expression set

```{r}
pr_rmv = rownames((Biobase::exprs(palmieri_eset_norm))) %in% probes
table(pr_rmv)
pr_rmv
```


```{r}
#need to keep the false values and remove the true ones
pr_rmv=!pr_rmv
table(pr_rmv)
palmieri_manfiltered <- subset(palmieri_eset_norm, pr_rmv)
head(Biobase::exprs(palmieri_manfiltered))
```

Next, the authors log-average the triplicate data into one value. We cannot average the log value of the samples, we should average the original linear value then log it to get the correct average.

```{r}
palmieri_manfiltered2=palmieri_manfiltered
exp_filtered = Biobase::exprs(palmieri_manfiltered2)

triplicate_cell_lines = c("A2008", "A2780", "C13", "CH1", "DOV13", "DOV13B", "FU-OV-1", "HeyA8", "HeyC2", "IGROV-1", "OV90", "OVCA420", "OVCA429", "OVCA433", "OVCAR-10", "OVCAR-8", "PA-1", "TYK-nu" )

rmvsamples=vector()

for(value in triplicate_cell_lines){

cell_lines=subset(SDRF2, FactorValue..CELL.LINE. == value)

cisplatin=subset(cell_lines, FactorValue..TREATMENT. == "cisplatin")
cisn=row.names(cisplatin)

cis_df = exp_filtered[,c(cisn)]
cisrown=row.names(cis_df)
cis_df=2^(cis_df)
rmeans=rowMeans(cis_df)
cis_df=cbind(cis_df, rmeans)
cis_df[,"rmeans"]=log2(cis_df[,"rmeans"])
cis_df[,1]=cis_df[,"rmeans"]
ciscoln=colnames(cis_df)
avgsamplen=(ciscoln[1])
avgsample=cis_df[,1, drop=FALSE]
avgsamplen
Biobase::exprs(palmieri_manfiltered2)[,avgsamplen]=avgsample
rmvsamples=c(rmvsamples, ciscoln[2:(length(ciscoln)-1)])

control=subset(cell_lines, FactorValue..TREATMENT. == "none")
contrn=row.names(control)

contr_df = exp_filtered[,c(contrn)]
contrown=row.names(contr_df)
contr_df=2^(contr_df)
rmeans=rowMeans(cis_df)
contr_df=cbind(contr_df, rmeans)
contr_df[,"rmeans"]=log2(contr_df[,"rmeans"])
contr_df[,1]=contr_df[,"rmeans"]
contrcoln=colnames(contr_df)
avgsamplen=(contrcoln[1])
avgsample=contr_df[,1, drop=FALSE]
avgsamplen
Biobase::exprs(palmieri_manfiltered2)[,avgsamplen]=avgsample
rmvsamples=c(rmvsamples, contrcoln[2:(length(contrcoln)-1)]) }

rmvsamples=c(rmvsamples, "GSM1160879_SMhu172.CEL")

#palmieri_manfiltered2 is the expression set with the averaged probes, still need to remove the rmvsamples from the expression set

rmvsamples

```


```{r}
totalsamples=colnames(Biobase::exprs(palmieri_manfiltered2))
keepsamples = totalsamples[! totalsamples %in% rmvsamples]
keepsamples
```


```{r}
palmieri_f = palmieri_manfiltered2[,keepsamples]
head(Biobase::exprs(palmieri_f))
```

```{r}
Biobase::pData(palmieri_f)
```


#### Annotation of the transcript clusters

```{r}
anno_palmieri <- AnnotationDbi::select(hugene10sttranscriptcluster.db,
                                  keys = (featureNames(palmieri_f)),
                                  columns = c("SYMBOL", "GENENAME"),
                                  keytype = "PROBEID")

anno_palmieri <- subset(anno_palmieri, !is.na(SYMBOL))

```

```{r}
anno_grouped <- group_by(anno_palmieri, PROBEID)
anno_summarized <- 
  dplyr::summarize(anno_grouped, no_of_matches = n_distinct(SYMBOL))

head(anno_summarized)
```

```{r}
anno_filtered <- filter(anno_summarized, no_of_matches > 1)

head(anno_filtered)
```

```{r}
probe_stats <- anno_filtered 

nrow(probe_stats)
```

1682 probes map to multiple genes

```{r}
ids_to_exlude <- (featureNames(palmieri_f) %in% probe_stats$PROBEID)
# returns a true or false vector based on whether the probes have multiple matches or not
# ids to exclude are those that return true
table(ids_to_exlude)
```

```{r}
palmieri_final <- subset(palmieri_f, !ids_to_exlude)

validObject(palmieri_final)
```

As we have just excluded probe IDs from the assay data, we now have to also exclude them from the feature data anno_palmieri:

```{r}
head(anno_palmieri)
```

Recall that fData enables us to access the feature data of an expression set. Until now, no feature data whatsoever is stored in the fData(palmieri_final). Only the row names are the row names of the assay data by default, which are the PROBEIDs of the transcripts.  

Therefore, we generate a column PROBEID in fData(palmieri_final) and assign the row names of fData(palmieri_final) to it:  

```{r}
fData(palmieri_final)$PROBEID <- rownames(fData(palmieri_final))
```

Then, we left-join fData(palmieri_final)with anno_palmieri, which already contains the columns “SYMBOL” and “GENENAME”. A left-join keeps the rows and columns of the first argument and adds the corresponding column entries of the second argument:

```{r}
fData(palmieri_final) <- left_join(fData(palmieri_final), anno_palmieri)

```

```{r}
#restore rownames after left_join
rownames(fData(palmieri_final)) <- fData(palmieri_final)$PROBEID 
    
validObject(palmieri_final)
```

#### Getting the cisplatin-induced response of both types of cells

To detect differential responses to cisplatin between epithelial and mesenchymal-like cell lines, the transcriptomic responses to cisplatin were computed by subtracting the gene expression value of control (untreated) from that of cisplatin-treated cells.

```{r}
SDRF_f = Biobase::pData(palmieri_final)
expr_f = Biobase::exprs(palmieri_final)
head(SDRF_f)
```

```{r}
head(expr_f)
```


```{r}
palmieri_diff = palmieri_final
SDRF_f = Biobase::pData(palmieri_diff)
expr_f = Biobase::exprs(palmieri_diff)

cell_line_names = unique(SDRF_f[,"FactorValue..CELL.LINE."])
length(cell_line_names)
```
```{r}

keep=vector()
for(value in cell_line_names){
cell_lines=subset(SDRF_f, FactorValue..CELL.LINE. == value)

cisplatin=subset(cell_lines, FactorValue..TREATMENT. == "cisplatin")
cisn=rownames(cisplatin)
cis_df = expr_f[,c(cisn), drop=FALSE]

control=subset(cell_lines, FactorValue..TREATMENT. == "none")
contrn=row.names(control)
contr_df = expr_f[,c(contrn), drop=FALSE]

cis_df=cis_df-contr_df
Biobase::exprs(palmieri_diff)[,cisn]=cis_df

keep=c(keep, cisn)}

length(keep)
```

```{r}
palmieri_d = palmieri_diff[,keep]
palmieri_d
```

palmieri_d is an expression set that contains only the difference in expression sets for each of the 46 cell lines.

```{r}
SDRF_f = Biobase::pData(palmieri_d)
expr_f = Biobase::exprs(palmieri_d)
```

#### Statistical Analysis



